File: libs/east/src/expr/integer.ts
/**
 * Copyright (c) 2025 Elara AI Pty Ltd
 * Dual-licensed under AGPL-3.0 and commercial license. See LICENSE for details.
 */
import type { AST } from "../ast.js";
import { get_location_id } from "../location.js";
import { FloatType, IntegerType, isTypeEqual } from "../types.js";
import { AstSymbol, Expr, FactorySymbol, type ToExpr } from "./expr.js";
import { valueOrExprToAst } from "./ast.js";
import type { FloatExpr } from "./float.js";
import type { BooleanExpr } from "./boolean.js";
import { equal, notEqual, less, lessEqual, greater, greaterEqual } from "./block.js";

/**
 * Expression representing integer values and operations.
 *
 * IntegerExpr provides arithmetic, comparison, and conversion operations for 64-bit signed integers.
 * Operations automatically promote to FloatExpr when mixed with float operands.
 *
 * @example
 * ```ts
 * // Basic arithmetic
 * const calculate = East.function([IntegerType, IntegerType], IntegerType, ($, a, b) => {
 *   const sum = a.add(b);
 *   const product = a.multiply(b);
 *   const power = a.pow(2n);
 *   $.return(sum.add(product).add(power));
 * });
 *
 * // Integer division and remainder
 * const divMod = East.function([IntegerType, IntegerType], StructType({quot: IntegerType, rem: IntegerType}), ($, x, y) => {
 *   $.return({quot: x.divide(y), rem: x.remainder(y)});
 * });
 *
 * // Type promotion with floats
 * const mixed = East.function([IntegerType], FloatType, ($, x) => {
 *   $.return(x.add(2.5));  // Returns FloatExpr
 * });
 * ```
 */
export class IntegerExpr extends Expr<IntegerType> {
  constructor(ast: AST, createExpr: ToExpr) {
    super(IntegerType, ast, createExpr);
  }

  /**
   * Negates an integer value (-x).
   *
   * @returns The negated integer
   *
   * @example
   * ```ts
   * const negate = East.function([IntegerType], IntegerType, ($, x) => {
   *   $.return(x.negate());
   * });
   * const compiled = East.compile(negate.toIR(), []);
   * compiled(5n);   // -5n
   * compiled(-3n);  // 3n
   * ```
   */
  negate(): IntegerExpr {
    return this[FactorySymbol]({
      ast_type: "Builtin",
      type: IntegerType,
      loc_id: get_location_id(),
      builtin: "IntegerNegate",
      type_parameters: [],
      arguments: [this[AstSymbol]],
    }) as IntegerExpr;
  }

  /**
   * Adds two integers or promotes to float if adding a float.
   *
   * @param y - The integer or float to add
   * @returns IntegerExpr if both operands are integers, FloatExpr if either is a float
   *
   * @example
   * ```ts
   * const addIntegers = East.function([IntegerType, IntegerType], IntegerType, ($, x, y) => {
   *   $.return(x.add(y));
   * });
   * const compiled = East.compile(addIntegers.toIR(), []);
   * compiled(3n, 4n);  // 7n
   *
   * // Adding with float promotes to float
   * const addFloat = East.function([IntegerType], FloatType, ($, x) => {
   *   $.return(x.add(2.5));
   * });
   * ```
   */
  add(y: Expr<IntegerType> | bigint): IntegerExpr
  add(y: Expr<FloatType> | number): FloatExpr
  add(y: any): Expr {
    const rightAst = valueOrExprToAst(y);
    if (isTypeEqual(rightAst.type, FloatType)) {
      return this[FactorySymbol]({
        ast_type: "Builtin",
        type: FloatType,
        loc_id: get_location_id(),
        builtin: "FloatAdd",
        type_parameters: [],
        arguments: [this.toFloat()[AstSymbol], rightAst],
      }) as FloatExpr;
    } else {
      return this[FactorySymbol]({
        ast_type: "Builtin",
        type: IntegerType,
        loc_id: get_location_id(),
        builtin: "IntegerAdd",
        type_parameters: [],
        arguments: [this[AstSymbol], rightAst],
      }) as IntegerExpr;
    }
  }

  /**
   * Subtracts two integers or promotes to float if subtracting a float.
   *
   * @param y - The integer or float to subtract
   * @returns IntegerExpr if both operands are integers, FloatExpr if either is a float
   *
   * @example
   * ```ts
   * const subtractIntegers = East.function([IntegerType, IntegerType], IntegerType, ($, x, y) => {
   *   $.return(x.subtract(y));
   * });
   * const compiled = East.compile(subtractIntegers.toIR(), []);
   * compiled(10n, 3n);  // 7n
   * compiled(5n, 8n);   // -3n
   * ```
   */
  subtract(y: Expr<IntegerType> | bigint): IntegerExpr
  subtract(y: Expr<FloatType> | number): FloatExpr
  subtract(y: any): Expr {
    const rightAst = valueOrExprToAst(y);
    if (isTypeEqual(rightAst.type, FloatType)) {
      return this[FactorySymbol]({
        ast_type: "Builtin",
        type: FloatType,
        loc_id: get_location_id(),
        builtin: "FloatSubtract",
        type_parameters: [],
        arguments: [this.toFloat()[AstSymbol], rightAst],
      }) as FloatExpr;
    } else {
      return this[FactorySymbol]({
        ast_type: "Builtin",
        type: IntegerType,
        loc_id: get_location_id(),
        builtin: "IntegerSubtract",
        type_parameters: [],
        arguments: [this[AstSymbol], rightAst],
      }) as IntegerExpr;
    }
  }

  /**
   * Multiplies two integers or promotes to float if multiplying by a float.
   *
   * @param y - The integer or float to multiply by
   * @returns IntegerExpr if both operands are integers, FloatExpr if either is a float
   *
   * @example
   * ```ts
   * const multiplyIntegers = East.function([IntegerType, IntegerType], IntegerType, ($, x, y) => {
   *   $.return(x.multiply(y));
   * });
   * const compiled = East.compile(multiplyIntegers.toIR(), []);
   * compiled(6n, 7n);   // 42n
   * compiled(-3n, 4n);  // -12n
   * ```
   */
  multiply(y: Expr<IntegerType> | bigint): IntegerExpr
  multiply(y: Expr<FloatType> | number): FloatExpr
  multiply(y: any): Expr {
    const rightAst = valueOrExprToAst(y);
    if (isTypeEqual(rightAst.type, FloatType)) {
      return this[FactorySymbol]({
        ast_type: "Builtin",
        type: FloatType,
        loc_id: get_location_id(),
        builtin: "FloatMultiply",
        type_parameters: [],
        arguments: [this.toFloat()[AstSymbol], rightAst],
      }) as FloatExpr;
    } else {
      return this[FactorySymbol]({
        ast_type: "Builtin",
        type: IntegerType,
        loc_id: get_location_id(),
        builtin: "IntegerMultiply",
        type_parameters: [],
        arguments: [this[AstSymbol], rightAst],
      }) as IntegerExpr;
    }
  }

  /**
   * Divides two integers (truncating division) or promotes to float if dividing by a float.
   *
   * @param y - The integer or float divisor
   * @returns IntegerExpr (quotient truncated toward zero) if both are integers, FloatExpr if either is a float
   *
   * @remarks When dividing by a float, uses normal floating point division
   *          (IEEE 754: division by zero yields Infinity/NaN, never throws).
   *
   * @throws East runtime error if the divisor is an integer zero.
   *
   * @example
   * ```ts
   * const divideIntegers = East.function([IntegerType, IntegerType], IntegerType, ($, x, y) => {
   *   $.return(x.divide(y));
   * });
   * const compiled = East.compile(divideIntegers.toIR(), []);
   * compiled(10n, 3n);  // 3n
   * compiled(10n, 2n);  // 5n
   * compiled(-10n, 3n); // -3n (truncated toward zero)
   * compiled(10n, 0n);  // throws East runtime error: Division by zero
   * ```
   */
  divide(y: Expr<IntegerType> | bigint): IntegerExpr
  divide(y: Expr<FloatType> | number): FloatExpr
  divide(y: any): Expr {
    const rightAst = valueOrExprToAst(y);
    if (isTypeEqual(rightAst.type, FloatType)) {
      return this[FactorySymbol]({
        ast_type: "Builtin",
        type: FloatType,
        loc_id: get_location_id(),
        builtin: "FloatDivide",
        type_parameters: [],
        arguments: [this.toFloat()[AstSymbol], rightAst],
      }) as FloatExpr;
    } else {
      return this[FactorySymbol]({
        ast_type: "Builtin",
        type: IntegerType,
        loc_id: get_location_id(),
        builtin: "IntegerDivide",
        type_parameters: [],
        arguments: [this[AstSymbol], rightAst],
      }) as IntegerExpr;
    }
  }

  /**
   * Computes the remainder of integer division or promotes to float.
   *
   * @param y - The integer or float divisor
   * @returns IntegerExpr if both are integers, FloatExpr if either is a float
   *
   * @remarks Result has the same sign as the dividend (truncated division semantics).
   *          When the divisor is a float, uses IEEE 754 remainder
   *          (remainder by float zero yields NaN, never throws).
   *
   * @throws East runtime error if the divisor is an integer zero.
   *
   * @example
   * ```ts
   * const getRemainder = East.function([IntegerType, IntegerType], IntegerType, ($, x, y) => {
   *   $.return(x.remainder(y));
   * });
   * const compiled = East.compile(getRemainder.toIR(), []);
   * compiled(10n, 3n);   // 1n (10 = 3*3 + 1)
   * compiled(10n, 4n);   // 2n
   * compiled(-10n, 3n);  // -1n (sign of the dividend)
   * compiled(10n, 0n);   // throws East runtime error: Division by zero
   * ```
   */
  remainder(y: Expr<IntegerType> | bigint): IntegerExpr
  remainder(y: Expr<FloatType> | number): FloatExpr
  remainder(y: any): Expr {
    const rightAst = valueOrExprToAst(y);
    if (isTypeEqual(rightAst.type, FloatType)) {
      return this[FactorySymbol]({
        ast_type: "Builtin",
        type: FloatType,
        loc_id: get_location_id(),
        builtin: "FloatRemainder",
        type_parameters: [],
        arguments: [this.toFloat()[AstSymbol], rightAst],
      }) as FloatExpr;
    } else {
      return this[FactorySymbol]({
        ast_type: "Builtin",
        type: IntegerType,
        loc_id: get_location_id(),
        builtin: "IntegerRemainder",
        type_parameters: [],
        arguments: [this[AstSymbol], rightAst],
      }) as IntegerExpr;
    }
  }

  /**
   * Raises an integer to a power or promotes to float.
   *
   * @param y - The integer or float exponent
   * @returns IntegerExpr if both are integers, FloatExpr if either is a float
   *
   * @remarks Integer exponentiation with negative exponent returns 0.
   *          When using a float exponent, uses floating point exponentiation.
   *
   * @example
   * ```ts
   * const power = East.function([IntegerType, IntegerType], IntegerType, ($, base, exp) => {
   *   $.return(base.pow(exp));
   * });
   * const compiled = East.compile(power.toIR(), []);
   * compiled(2n, 3n);    // 8n (2^3)
   * compiled(5n, 2n);    // 25n
   * compiled(10n, 0n);   // 1n (anything^0 = 1)
   * compiled(2n, -1n);   // 0n (negative exponent returns 0 for integers)
   * ```
   */
  pow(y: Expr<IntegerType> | bigint): IntegerExpr
  pow(y: Expr<FloatType> | number): FloatExpr
  pow(y: any): Expr {
    const rightAst = valueOrExprToAst(y);
    if (isTypeEqual(rightAst.type, FloatType)) {
      return this[FactorySymbol]({
        ast_type: "Builtin",
        type: FloatType,
        loc_id: get_location_id(),
        builtin: "FloatPow",
        type_parameters: [],
        arguments: [this.toFloat()[AstSymbol], rightAst],
      }) as FloatExpr;
    } else {
      return this[FactorySymbol]({
        ast_type: "Builtin",
        type: IntegerType,
        loc_id: get_location_id(),
        builtin: "IntegerPow",
        type_parameters: [],
        arguments: [this[AstSymbol], rightAst],
      }) as IntegerExpr;
    }
  }

  /**
   * Returns the absolute value of an integer.
   *
   * @returns The absolute value (always non-negative)
   *
   * @example
   * ```ts
   * const absoluteValue = East.function([IntegerType], IntegerType, ($, x) => {
   *   $.return(x.abs());
   * });
   * const compiled = East.compile(absoluteValue.toIR(), []);
   * compiled(5n);    // 5n
   * compiled(-5n);   // 5n
   * compiled(0n);    // 0n
   * ```
   */
  abs(): IntegerExpr {
    return this[FactorySymbol]({
      ast_type: "Builtin",
      type: IntegerType,
      loc_id: get_location_id(),
      builtin: "IntegerAbs",
      type_parameters: [],
      arguments: [this[AstSymbol]],
    }) as IntegerExpr;
  }

  /**
   * Returns the sign of an integer (-1 for negative, 0 for zero, 1 for positive).
   *
   * @returns -1n, 0n, or 1n depending on the sign of the integer
   *
   * @example
   * ```ts
   * const getSign = East.function([IntegerType], IntegerType, ($, x) => {
   *   $.return(x.sign());
   * });
   * const compiled = East.compile(getSign.toIR(), []);
   * compiled(42n);   // 1n
   * compiled(-17n);  // -1n
   * compiled(0n);    // 0n
   * ```
   */
  sign(): IntegerExpr {
    return this[FactorySymbol]({
      ast_type: "Builtin",
      type: IntegerType,
      loc_id: get_location_id(),
      builtin: "IntegerSign",
      type_parameters: [],
      arguments: [this[AstSymbol]],
    }) as IntegerExpr;
  }

  /**
   * Computes the integer logarithm with a specified base (floor of log).
   *
   * @param base - The base for the logarithm
   * @returns The floor of log_base(x)
   *
   * @remarks Returns 0 if x <= 0 or base <= 1.
   *
   * @example
   * ```ts
   * const logarithm = East.function([IntegerType, IntegerType], IntegerType, ($, x, base) => {
   *   $.return(x.log(base));
   * });
   * const compiled = East.compile(logarithm.toIR(), []);
   * compiled(1000n, 10n);  // 3n (10^3 = 1000)
   * compiled(8n, 2n);      // 3n (2^3 = 8)
   * compiled(100n, 10n);   // 2n (10^2 = 100)
   * compiled(7n, 2n);      // 2n (floor of log_2(7) ≈ 2.8)
   * ```
   */
  log(base: Expr<IntegerType> | bigint): IntegerExpr {
    const baseAst = valueOrExprToAst(base);
    return this[FactorySymbol]({
      ast_type: "Builtin",
      type: IntegerType,
      loc_id: get_location_id(),
      builtin: "IntegerLog",
      type_parameters: [],
      arguments: [this[AstSymbol], baseAst],
    }) as IntegerExpr;
  }

  /**
   * Converts an integer to a floating-point number.
   *
   * @returns The integer as a FloatExpr
   *
   * @remarks May lose precision for very large integers (beyond 2^53).
   *
   * @example
   * ```ts
   * const convertToFloat = East.function([IntegerType], FloatType, ($, x) => {
   *   $.return(x.toFloat());
   * });
   * const compiled = East.compile(convertToFloat.toIR(), []);
   * compiled(42n);    // 42.0
   * compiled(-17n);   // -17.0
   * compiled(0n);     // 0.0
   * ```
   */
  toFloat(): FloatExpr {
    return this[FactorySymbol]({
      ast_type: "Builtin",
      type: FloatType,
      loc_id: get_location_id(),
      builtin: "IntegerToFloat",
      type_parameters: [],
      arguments: [this[AstSymbol]],
    }) as FloatExpr;
  }

  /**
   * Checks if this integer equals another value.
   *
   * @param other - The value to compare against
   * @returns A BooleanExpr that is true if the values are equal
   *
   * @example
   * ```ts
   * const isEqual = East.function([IntegerType, IntegerType], BooleanType, ($, a, b) => {
   *   $.return(a.equals(b));
   * });
   * const compiled = East.compile(isEqual.toIR(), []);
   * compiled(5n, 5n);   // true
   * compiled(5n, 3n);   // false
   * ```
   */
  equals(other: IntegerExpr | bigint): BooleanExpr {
    return equal(this, other);
  }

  /**
   * Checks if this integer does not equal another value.
   *
   * @param other - The value to compare against
   * @returns A BooleanExpr that is true if the values are not equal
   *
   * @example
   * ```ts
   * const isNotEqual = East.function([IntegerType, IntegerType], BooleanType, ($, a, b) => {
   *   $.return(a.notEquals(b));
   * });
   * const compiled = East.compile(isNotEqual.toIR(), []);
   * compiled(5n, 3n);   // true
   * compiled(5n, 5n);   // false
   * ```
   */
  notEquals(other: IntegerExpr | bigint): BooleanExpr {
    return notEqual(this, other);
  }

  /**
   * Checks if this integer is greater than another value.
   *
   * @param other - The value to compare against
   * @returns A BooleanExpr that is true if this value is greater
   *
   * @example
   * ```ts
   * const isGreater = East.function([IntegerType, IntegerType], BooleanType, ($, a, b) => {
   *   $.return(a.greaterThan(b));
   * });
   * const compiled = East.compile(isGreater.toIR(), []);
   * compiled(5n, 3n);   // true
   * compiled(3n, 5n);   // false
   * compiled(5n, 5n);   // false
   * ```
   */
  greaterThan(other: IntegerExpr | bigint): BooleanExpr {
    return greater(this, other);
  }

  /**
   * Checks if this integer is less than another value.
   *
   * @param other - The value to compare against
   * @returns A BooleanExpr that is true if this value is less
   *
   * @example
   * ```ts
   * const isLess = East.function([IntegerType, IntegerType], BooleanType, ($, a, b) => {
   *   $.return(a.lessThan(b));
   * });
   * const compiled = East.compile(isLess.toIR(), []);
   * compiled(3n, 5n);   // true
   * compiled(5n, 3n);   // false
   * compiled(5n, 5n);   // false
   * ```
   */
  lessThan(other: IntegerExpr | bigint): BooleanExpr {
    return less(this, other);
  }

  /**
   * Checks if this integer is greater than or equal to another value.
   *
   * @param other - The value to compare against
   * @returns A BooleanExpr that is true if this value is greater than or equal
   *
   * @example
   * ```ts
   * const isGreaterOrEqual = East.function([IntegerType, IntegerType], BooleanType, ($, a, b) => {
   *   $.return(a.greaterThanOrEqual(b));
   * });
   * const compiled = East.compile(isGreaterOrEqual.toIR(), []);
   * compiled(5n, 3n);   // true
   * compiled(5n, 5n);   // true
   * compiled(3n, 5n);   // false
   * ```
   */
  greaterThanOrEqual(other: IntegerExpr | bigint): BooleanExpr {
    return greaterEqual(this, other);
  }

  /**
   * Checks if this integer is less than or equal to another value.
   *
   * @param other - The value to compare against
   * @returns A BooleanExpr that is true if this value is less than or equal
   *
   * @example
   * ```ts
   * const isLessOrEqual = East.function([IntegerType, IntegerType], BooleanType, ($, a, b) => {
   *   $.return(a.lessThanOrEqual(b));
   * });
   * const compiled = East.compile(isLessOrEqual.toIR(), []);
   * compiled(3n, 5n);   // true
   * compiled(5n, 5n);   // true
   * compiled(5n, 3n);   // false
   * ```
   */
  lessThanOrEqual(other: IntegerExpr | bigint): BooleanExpr {
    return lessEqual(this, other);
  }

  // ============================================================================
  // Aliases for comparison operations
  // ============================================================================

  /** Alias for {@link equals} */
  eq = this.equals;
  /** Alias for {@link equals} */
  equal = this.equals;

  /** Alias for {@link notEquals} */
  ne = this.notEquals;
  /** Alias for {@link notEquals} */
  notEqual = this.notEquals;

  /** Alias for {@link greaterThan} */
  gt = this.greaterThan;
  /** Alias for {@link greaterThan} */
  greater = this.greaterThan;

  /** Alias for {@link lessThan} */
  lt = this.lessThan;
  /** Alias for {@link lessThan} */
  less = this.lessThan;

  /** Alias for {@link greaterThanOrEqual} */
  gte = this.greaterThanOrEqual;
  /** Alias for {@link greaterThanOrEqual} */
  ge = this.greaterThanOrEqual;
  /** Alias for {@link greaterThanOrEqual} */
  greaterEqual = this.greaterThanOrEqual;

  /** Alias for {@link lessThanOrEqual} */
  lte = this.lessThanOrEqual;
  /** Alias for {@link lessThanOrEqual} */
  le = this.lessThanOrEqual;
  /** Alias for {@link lessThanOrEqual} */
  lessEqual = this.lessThanOrEqual;

  // ============================================================================
  // Aliases for arithmetic operations
  // ============================================================================

  /** Alias for {@link add} */
  plus = this.add;

  /** Alias for {@link subtract} */
  sub = this.subtract;
  /** Alias for {@link subtract} */
  minus = this.subtract;

  /** Alias for {@link multiply} */
  mul = this.multiply;
  /** Alias for {@link multiply} */
  times = this.multiply;

  /** Alias for {@link divide} */
  div = this.divide;

  /** Alias for {@link remainder} */
  mod = this.remainder;
  /** Alias for {@link remainder} */
  rem = this.remainder;
  /** Alias for {@link remainder} */
  modulo = this.remainder;
}

File: libs/east-py/packages/east-py/east/expression/expr/integer.py
#
# Copyright (c) 2025 Elara AI Pty Ltd
# Licensed under the Business Source License 1.1. See LICENSE.md for details.
#
"""``IntegerExpression`` — TS ``IntegerExpr`` (``libs/east/src/expr/integer.ts``)."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from east.expression.errors import ExpressionError
from east.expression.expr.base import Expression
from east.expression.lift import _lift
from east.expression.nodes import _builtin
from east.types.types import FloatType, IntegerType

_FLOORDIV_FORK = (
    "python `//` floors the quotient; East IntegerDivide truncates toward "
    "zero (-10 // 3 is -4 in python but -3 traced) — call "
    "East.Integer.divide(a, b) explicitly"
)
_MOD_FORK = (
    "python `%` takes the sign of the divisor; East IntegerRemainder / "
    "FloatRemainder take the sign of the dividend (-10 % 3 is 2 in python "
    "but -1 traced) — call East.Integer.remainder(a, b) or "
    "East.Float.remainder(a, b) explicitly"
)
_POW_FORK = (
    "python `**` promotes a negative Integer exponent to float (2 ** -1 is "
    "0.5 in python but 0 as East IntegerPow) — call East.Integer.pow(a, b) "
    "explicitly"
)
_DIV_AMBIGUOUS = (
    "`/` on East Integers is ambiguous — call East.Integer.divide(a, b) "
    "for truncating integer division or .to_float() for float division"
)

if TYPE_CHECKING:
    from east.expression.expr.float import FloatExpression


class IntegerExpression(Expression):
    """64-bit signed Integer arithmetic and conversion. Operators overload
    only where python agrees with East (#624): ``+`` ``-`` ``*`` and unary
    ``-``; ``//``, ``%``, ``**`` and ``/`` raise naming the explicit East
    spelling that says which semantics you get."""

    __slots__ = ()
    _kind = "Integer"

    def _binary(self, name: str, other: Any, reflected: bool = False) -> IntegerExpression:
        other = _lift(other, hint=IntegerType)
        lhs, rhs = (other, self) if reflected else (self, other)
        if other.east_type.type != "Integer":
            raise ExpressionError(
                f"arithmetic between {lhs.east_type.type} and {rhs.east_type.type} — East "
                "has no implicit numeric coercion; convert explicitly with .to_float()"
            )
        return self._expr(_builtin(name, IntegerType, [], [lhs.ir, rhs.ir]), IntegerType)

    # ── arithmetic ─────────────────────────────────────────────────────

    def __add__(self, other: Any) -> IntegerExpression:
        return self._binary("IntegerAdd", other)

    def __radd__(self, other: Any) -> IntegerExpression:
        return self._binary("IntegerAdd", other, reflected=True)

    def __sub__(self, other: Any) -> IntegerExpression:
        return self._binary("IntegerSubtract", other)

    def __rsub__(self, other: Any) -> IntegerExpression:
        return self._binary("IntegerSubtract", other, reflected=True)

    def __mul__(self, other: Any) -> IntegerExpression:
        return self._binary("IntegerMultiply", other)

    def __rmul__(self, other: Any) -> IntegerExpression:
        return self._binary("IntegerMultiply", other, reflected=True)

    def __mod__(self, other: Any) -> IntegerExpression:
        raise ExpressionError(_MOD_FORK)

    def __rmod__(self, other: Any) -> IntegerExpression:
        raise ExpressionError(_MOD_FORK)

    def __pow__(self, other: Any) -> IntegerExpression:
        raise ExpressionError(_POW_FORK)

    def __rpow__(self, other: Any) -> IntegerExpression:
        raise ExpressionError(_POW_FORK)

    def __truediv__(self, other: Any) -> IntegerExpression:
        raise ExpressionError(_DIV_AMBIGUOUS)

    def __rtruediv__(self, other: Any) -> IntegerExpression:
        raise ExpressionError(_DIV_AMBIGUOUS)

    def __floordiv__(self, other: Any) -> IntegerExpression:
        raise ExpressionError(_FLOORDIV_FORK)

    def __rfloordiv__(self, other: Any) -> IntegerExpression:
        raise ExpressionError(_FLOORDIV_FORK)

    def __neg__(self) -> IntegerExpression:
        return self._expr(_builtin("IntegerNegate", IntegerType, [], [self.ir]), IntegerType)

    def __abs__(self) -> IntegerExpression:
        return self.abs()

    # ── conversions and math ────────────────────────────────────────────

    def to_float(self) -> FloatExpression:
        """Traced IntegerToFloat (TS ``toFloat``)."""
        return self._expr(_builtin("IntegerToFloat", FloatType, [], [self.ir]), FloatType)

    def abs(self) -> IntegerExpression:
        """Traced IntegerAbs."""
        return self._expr(_builtin("IntegerAbs", IntegerType, [], [self.ir]), IntegerType)

    def sign(self) -> IntegerExpression:
        """Traced IntegerSign: -1, 0 or 1."""
        return self._expr(_builtin("IntegerSign", IntegerType, [], [self.ir]), IntegerType)

    def log(self, base: Any) -> IntegerExpression:
        """Traced IntegerLog: the floor of ``log_base(x)`` (TS ``log(base)``)."""
        b = self._typed("log", base, IntegerType)
        return self._expr(_builtin("IntegerLog", IntegerType, [], [self.ir, b.ir]), IntegerType)

    # ── the named arithmetic (TS ``add`` … ``pow``): a Float operand promotes ──

    def _named(self, integer_op: str, float_op: str, other: Any) -> Any:
        """``self <op> other`` as TS spells it: an Integer operand keeps the
        Integer builtin; a Float operand converts ``self`` with IntegerToFloat
        first (the TypeScript ``IntegerExpr`` promotion)."""
        o = _lift(other, hint=IntegerType)
        if o.east_type.type == "Float":
            return self._expr(
                _builtin(float_op, FloatType, [], [self.to_float().ir, o.ir]), FloatType)
        if o.east_type.type != "Integer":
            raise ExpressionError(
                f"arithmetic between Integer and {o.east_type.type} — East has no "
                "implicit numeric coercion; convert explicitly with .to_float()")
        return self._expr(_builtin(integer_op, IntegerType, [], [self.ir, o.ir]), IntegerType)

    def negate(self) -> IntegerExpression:
        """Traced IntegerNegate (TS ``negate``; also unary ``-``)."""
        return -self

    def add(self, y: Any) -> Any:
        """Traced IntegerAdd — FloatAdd of ``to_float()`` for a Float ``y`` (TS ``add``)."""
        return self._named("IntegerAdd", "FloatAdd", y)

    def subtract(self, y: Any) -> Any:
        """Traced IntegerSubtract (TS ``subtract``)."""
        return self._named("IntegerSubtract", "FloatSubtract", y)

    def multiply(self, y: Any) -> Any:
        """Traced IntegerMultiply (TS ``multiply``)."""
        return self._named("IntegerMultiply", "FloatMultiply", y)

    def divide(self, y: Any) -> Any:
        """Traced IntegerDivide — truncating toward zero, an East runtime
        error on a zero divisor (TS ``divide``); FloatDivide for a Float ``y``."""
        return self._named("IntegerDivide", "FloatDivide", y)

    def remainder(self, y: Any) -> Any:
        """Traced IntegerRemainder — the sign of the dividend (TS ``remainder``)."""
        return self._named("IntegerRemainder", "FloatRemainder", y)

    def pow(self, y: Any) -> Any:
        """Traced IntegerPow — a negative exponent yields 0 (TS ``pow``)."""
        return self._named("IntegerPow", "FloatPow", y)

    # the TypeScript aliases
    plus = add
    sub = subtract
    minus = subtract
    mul = multiply
    times = multiply
    div = divide
    mod = remainder
    rem = remainder
    modulo = remainder
